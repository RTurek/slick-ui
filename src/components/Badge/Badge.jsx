import React from 'react';
import PropTypes from 'prop-types';
import MaterialBadge from '@material-ui/core/Badge';
import { useTheme, withStyles } from '@material-ui/core/styles';
import { COLORS } from '../../styles/colors';
import hexToRgba from '../../utils/hexToRgba';
import { decagramWithBorderEncoded, santaHatEncoded } from '../../assets/encodedSVGs';

function buildStyle(props) {
  const style = {
    lineHeight: 1,
    boxShadow: `0 2px 4px 0 ${hexToRgba(props.theme.palette.alert.main, 0.4)}`,
    color: props.theme.palette.common.white
  };

  if (props.customBadgeVariant === 'dot') {
    style.top = '0.2rem';
    style.right = '0.2rem';
    style.backgroundColor = props.theme.palette.alert.main;
  } else if (props.customBadgeVariant === 'menu-dot') {
    style.top = '0.7rem';
    style.right = '0.7rem';
    style.backgroundColor = props.theme.palette.alert.main;
  } else if (props.customBadgeVariant === 'mobile-dot') {
    style.top = '0.9rem';
    style.right = '35%';
    style.backgroundColor = props.theme.palette.alert.main;
  } else if (props.customBadgeVariant === 'alert-right') {
    style.top = '0.3rem';
    style.right = '0';
    style.backgroundColor = props.theme.palette.alert.main;
  } else if (props.customBadgeVariant === 'alert-top-right') {
    style.top = '-10px';
    style.right = '0.8rem';
    style.fontSize = '0.7rem';
    style.backgroundColor = props.theme.palette.alert.main;
  } else if (props.customBadgeVariant === 'point-progress') {
    style.top = '-5px';
    style.right = '50px';
    style.backgroundColor = props.theme.palette.secondary.main;
    style.boxShadow = `0 2px 4px 0 ${hexToRgba(props.theme.palette.secondary.main, 0.4)}`;
  } else if (props.customBadgeVariant === 'grey-left') {
    style.top = '0rem';
    style.right = '2rem';
    style.height = '22px';
    style.width = '22px';
    style.boxShadow = 'none';
    style.borderRadius = '50%';
    style.fontSize = '0.8rem';
    style.backgroundColor = props.theme.palette.mediumGrey.main;
  } else if (props.customBadgeVariant === 'decagram-left') {
    style.top = '0.20rem';
    style.right = '1.80rem';
    style.height = '26px';
    style.width = '27px';
    style.boxShadow = 'none';
    style.borderRadius = '50%';
    style.fontSize = '0.8rem';
    style.backgroundImage = `url("data:image/svg+xml;base64,${decagramWithBorderEncoded}")`;
  } else if (props.customBadgeVariant === 'seasonal-badge-left') {
    style.left = '-1.8rem';
    style.top = '0rem';
    style.width = '32px';
    style.height = '28px';
    style.borderRadius = '0px';
    style.padding = '0px';
    style.margin = '0px';
    style.boxShadow = 'none';
    style.color = props.theme.palette.common.black;
    style.backgroundImage = `url("data:image/svg+xml;base64,${santaHatEncoded}")`;
  } else if (props.customBadgeVariant === 'bottom-right') {
    style.right = '0';
    style.bottom = '0';
    style.top = 'unset !important';
    style.backgroundColor = props.theme.palette.warning.main;
    style.transform = 'scale(1) translateX(50%)';
  }

  return {
    ...style,
    ...props.style
  };
}

function Badge(props) {
  const styles = {
    badge: buildStyle({
      ...props,
      theme: useTheme()
    })
  };

  const VariantBadge = withStyles(styles)((variantProps) => {
    const {
      classes,
      children,
      style,
      customBadgeVariant,
      ...otherProps
    } = variantProps;

    return (
      <MaterialBadge
        variant={customBadgeVariant.includes('dot') ? 'dot' : 'standard'}
        classes={{
          badge: classes.badge
        }}
        {...otherProps}
      >
        {children}
      </MaterialBadge>
    );
  });

  return (
    <VariantBadge {...props} />
  );
}

Badge.propTypes = {
  /** customBadgeVariant of various types */
  customBadgeVariant: PropTypes.oneOf(['alert-right', 'grey-left', 'decagram-left', 'menu-dot', 'dot', 'point-progress', 'bottom-right', 'alert-top-right', 'seasonal-badge-left']),
  /** invisible true/false */
  invisible: PropTypes.bool,
  /** badgeContent typically a numerical value representing # of notifications, etc. */
  badgeContent: PropTypes.node,
  /** max number to show in badge */
  max: PropTypes.number,
  /** theming object */
  theme: PropTypes.shape({
    palette: PropTypes.shape({
      mediumGrey: PropTypes.shape({
        main: PropTypes.string
      }),
      alert: PropTypes.shape({
        main: PropTypes.string
      }),
      warning: PropTypes.shape({
        main: PropTypes.string
      }),
      secondary: PropTypes.shape({
        main: PropTypes.string
      }),
      text: PropTypes.shape({
        secondary: PropTypes.string
      }),
      common: PropTypes.shape({
        white: PropTypes.string
      })
    })
  })
};

Badge.defaultProps = {
  invisible: undefined,
  badgeContent: null,
  customBadgeVariant: 'alert-right',
  max: 9,
  theme: {
    palette: {
      mediumGrey: {
        main: COLORS.mediumGrey
      },
      alert: {
        main: COLORS.alert
      },
      warning: {
        main: COLORS.warning
      },
      secondary: {
        main: COLORS.secondary
      },
      text: {
        secondary: COLORS.gray
      },
      common: {
        white: COLORS.binary,
        black: COLORS.black
      }
    }
  }
};

export default Badge;
/* For Storybook Use Only */
export const BadgeForStorybook = Badge;
