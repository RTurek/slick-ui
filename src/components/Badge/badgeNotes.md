#### Badge
---
This variant on the Material Badge component gives some opinionated options for color and placement of a few common badge use cases. Just wrap this around whatever child component(s) you want!

#### Notes
---
Yes, that is a Santa hat. You're welcome ;P (seasonalBadgeLeft)

#### Sources
---
- https://mui.com/components/badges/
