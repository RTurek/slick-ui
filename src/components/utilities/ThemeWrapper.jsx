import React from 'react';
import PropTypes from 'prop-types';
import { createTheme } from '@material-ui/core/styles';
import { ThemeProvider } from '@material-ui/styles';
import { COLORS } from '@styles/colors';

class ThemeWrapper extends React.Component {
  constructor(props) {
    super(props);

    const primaryThemeColor = COLORS.primary
    const secondaryThemeColor = COLORS.secondary
    const textPrimaryThemeColor = COLORS.textPrimary;
    const textSecondaryThemeColor = COLORS.textSecondary;
    const backgroundThemeColor = COLORS.background;
    const alertThemeColor = COLORS.alert;
    const urgentThemeColor = COLORS.urgent;
    const successThemeColor = COLORS.success;
    const warningThemeColor = COLORS.warning;
    const crGreenThemeColor = COLORS.crGreen;
    const rpOrangeThemeColor = COLORS.rpOrange;
    const darkGreyThemeColor = COLORS.darkGrey;
    const mediumGreyThemeColor = COLORS.mediumGrey;
    const lightGreyThemeColor = COLORS.lightGrey;
    const modernBlackThemeColor = COLORS.modernBlack;

    this.muiTheme = createTheme({
      typography: {
        // fontFamily: primaryFontFamily,
        useNextVariants: true,
        h1: {
          letterSpacing: '-1.5px'
        },
        h2: {
          letterSpacing: '-0.5px'
        },
        h3: {
          lineHeight: 1,
          letterSpacing: '0px',
          fontWeight: '300'
        },
        h4: {
          fontSize: '1.875rem',
          letterSpacing: '0.25px',
          lineHeight: 1.0667
        },
        h5: {
          letterSpacing: '0px',
          lineHeight: 1.0833
        },
        h6: {
          letterSpacing: '0.15px',
          lineHeight: 1.1
        },
        subtitle1: {
          letterSpacing: '0.15px',
          lineHeight: 1.125
        },
        subtitle2: {
          letterSpacing: '0.1px',
          lineHeight: 1.1428
        },
        body1: {
          letterSpacing: '0.25px',
          lineHeight: 1.25
        },
        body2: {
          letterSpacing: '0.15px',
          lineHeight: 1.28571
        },
        body1Next: {
          letterSpacing: '0.25px',
          lineHeight: 1.25
        },
        body2Next: {
          letterSpacing: '0.15px',
          lineHeight: 1.28571
        },
        button: {
          fontWeight: 700,
          letterSpacing: '1.25px'
        },
        buttonNext: {
          fontWeight: 700,
          letterSpacing: '1.25px'
        },
        caption: {
          letterSpacing: '0.2px',
          lineHeight: 1.33333
        },
        overline: {
          fontWeight: 600,
          fontSize: '0.625rem',
          letterSpacing: '1.5px',
          lineHeight: 1.4
        },
        fontWeightLight: 300,
        fontWeightMedium: 600,
        fontWeightRegular: 400
      },
      palette: {
        primary: {
          main: primaryThemeColor
        },
        secondary: {
          main: secondaryThemeColor,
          contrastText: '#FFF'
        },
        error: {
          main: alertThemeColor
        },
        text: {
          primary: textPrimaryThemeColor,
          secondary: textSecondaryThemeColor
        },
        alert: {
          main: alertThemeColor
        },
        urgent: {
          main: urgentThemeColor
        },
        success: {
          main: successThemeColor
        },
        warning: {
          main: warningThemeColor
        },
        backgroundColor: {
          main: backgroundThemeColor
        },
        mediumGrey: {
          main: mediumGreyThemeColor
        },
        lightGrey: {
          main: lightGreyThemeColor
        },
        crGreen: {
          main: crGreenThemeColor
        },
        darkGrey: {
          main: darkGreyThemeColor
        },
        rpOrange: {
          main: rpOrangeThemeColor
        },
        modernBlack: {
          main: modernBlackThemeColor
        }
      },
      overrides: {
        MuiTypography: {
          root: {
            display: 'block'
          }
        }
      },
      props: {
        MuiButtonBase: {
          TouchRippleProps: {
            style: {
              color: lightGreyThemeColor
            }
          }
        }
      }
    });
  }

  render() {
    return (
      <ThemeProvider theme={this.muiTheme}>
        {this.props.children}
      </ThemeProvider>
    );
  }
}

ThemeWrapper.propTypes = {
  children: PropTypes.node
};

ThemeWrapper.defaultProps = {
  children: null
};

export default ThemeWrapper;
