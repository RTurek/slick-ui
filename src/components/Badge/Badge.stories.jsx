import React from 'react';
import { withInfo } from '@storybook/addon-info';
import AccountCircle from '@material-ui/icons/AccountCircle';
import { storybookStyles } from '../../../.storybook/storybookTheming';
import StoryWrapper from '../utilities/StoryWrapper';
import Badge, { BadgeForStorybook } from './Badge';
import badgeNotes from './badgeNotes.md';

export default {
  title: 'Core / Badge',

  decorators: [
    withInfo({
      text: badgeNotes,
      header: true,
      inline: true,
      source: false,
      styles: storybookStyles,
      propTables: [BadgeForStorybook],
      propTablesExclude: [Badge, AccountCircle],
      excludedPropTypes: ['theme']
    })
  ]
};

export const alertRight = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent={1} customBadgeVariant="alert-right">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

alertRight.story = {
  name: 'Alert Right'
};

export const alertText = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent="NEW" customBadgeVariant="alert-right">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

alertText.story = {
  name: 'Alert Text'
};

export const alertDot = () => (
  <StoryWrapper>
    <Badge invisible={false} customBadgeVariant="dot">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

alertDot.story = {
  name: 'Alert Dot'
};

export const greyLeft = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent={3} customBadgeVariant="grey-left">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

greyLeft.story = {
  name: 'Grey Left'
};

export const decagramLeft = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent={3} customBadgeVariant="decagram-left">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

decagramLeft.story = {
  name: 'Decagram Left'
};

export const seasonalBadgeLeft = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent={3} customBadgeVariant="seasonal-badge-left">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

seasonalBadgeLeft.story = {
  name: 'Seasonal Badge Left'
};

export const bottomRight = () => (
  <StoryWrapper>
    <Badge invisible={false} badgeContent={3} customBadgeVariant="bottom-right">
      <AccountCircle fontSize="large" />
    </Badge>
  </StoryWrapper>
);

bottomRight.story = {
  name: 'Bottom Right'
};
