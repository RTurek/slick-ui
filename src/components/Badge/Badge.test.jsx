import React from 'react';
import renderer from 'react-test-renderer';
import AccountCircle from '@material-ui/icons/AccountCircle';
import Badge from './Badge';
import ThemeWrapper from '../utilities/ThemeWrapper';

test('Badge should render', () => {
  const component = (
    <ThemeWrapper>
      <Badge invisible={false} badgeContent={1} customBadgeVariant="alert-right">
        <AccountCircle fontSize="large" />
      </Badge>
    </ThemeWrapper>
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
