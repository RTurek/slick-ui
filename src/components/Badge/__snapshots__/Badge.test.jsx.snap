// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render 1`] = `
<span
  className="MuiBadge-root"
  theme={
    Object {
      "palette": Object {
        "alert": Object {
          "main": "#FF3D7F",
        },
        "common": Object {
          "black": undefined,
          "white": "#FFFFFF",
        },
        "mediumGrey": Object {
          "main": "#7E8E9F",
        },
        "secondary": Object {
          "main": "#00A0DF",
        },
        "text": Object {
          "secondary": "#929292",
        },
        "warning": Object {
          "main": "#FB8B23",
        },
      },
    }
  }
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
    />
  </svg>
  <span
    className="MuiBadge-badge Component-badge-1 MuiBadge-anchorOriginTopRightRectangle"
  >
    1
  </span>
</span>
`;
